### ISSUE TITLE
[Action Graph] Add Persistent Variables

### ISSUE THREAD

**Issue by Nolankicks** [Facepunch/sbox-issues#7284]
### What can't you do?

I would like to be able to define persistent variables I would be able to use in different graphs.

### How would you like it to work?

I’m not sure the best implementation of this. So please leave suggestions in the comments. But some sort of graph referencing would be good. 

### What have you tried?

There is no way to do this out of the box.

**Comment by CarsonKompon** [comment]
Is this basically what you're asking for?
https://docs.facepunch.com/s/sbox-dev/doc/component-editor-1YIBDlDz1W

**Comment by Nolankicks** [comment]
Nope, I want to define them per graph. Like public GameObject gameObject { get; set; }. I don't want components, I want to just create variables I can just and reference like c# within the graph and across graphs.

**Comment by MrBrax** [comment]
variables would be stored internally on the component that the action belongs to, no?

**Comment by Nolankicks** [comment]
Ok let's limit this scope to per graph variables that you can create, to store values etc. The way it is setup now is very hard to use and not intuitive.

**Comment by CarsonKompon** [comment]
> variables would be stored internally on the component that the action belongs to, no?

Something like this could work. I think there just needs to be a good use case.
Do other engines even support this? If so how


> Ok let's limit this scope to per graph variables that you can create, to store values etc. The way it is setup now is very hard to use and not intuitive.

That sounds much more reasonable, has a lot of use cases I can imagine.

**Comment by MrBrax** [comment]
my use case a couple months ago was making a simple fire alarm beep every ten seconds, i wanted to use timesince to delay it, since i couldn't make a while loop with delays in action graph i'm pretty sure.

for this extremely simple thing i instead had to make a c# component of like ten lines, which seems a bit extra for a tiny aesthetic feature

**Comment by MrBrax** [comment]
those won't persist between runs though, right? it will complain about the variable not being defined if you try to check it first, and i couldn't find a way to set it if it was null because it always errored

**Comment by Nolankicks** [comment]
> those won't persist between runs though, right? it will complain about the variable not being defined if you try to check it first, and i couldn't find a way to set it if it was null because it always errored

Yes, I can deal with the poor interface if they can be set persistently. For example, there could be some kind of menu where you can press a button to add a new variable, set the name, type, and starting value, and use those within the graph and the value would not be reset each time the graph is run. There could be a setting to reset the value each time the graph is run.

**Comment by CarsonKompon** [comment]
Yeah looks like the value needs to be assigned to before it can be accessed, so it's currently not persistent. I think it's best to let @Metapyziks decide how he wants to handle this since he knows best.

Either way I'll improve the workflow of creating the current actiongraph variables since it's a bit hidden away.

**Comment by Metapyziks** [comment]
> Yes, I can deal with the poor interface if they can be set persistently. For example, there could be some kind of menu where you can press a button to add a new variable, set the name, type, and starting value, and use those within the graph and the value would not be reset each time the graph is run. This panel would be like the inspect panel in the top left (default layout). There could be a setting to reset the value each time the graph is run. [@CarsonKompon](https://github.com/CarsonKompon)

This is almost exactly what you can do with the component editor library Carson linked, just you group properties into components so you can access them from more than one graph.

I don't like the idea of having properties local to a graph because you'll eventually want to be able to get / set them from other graphs. Let's just make the component editor / a simpler version of it built in.

**Comment by Nolankicks** [comment]
I don't love having to make components just to store variables, but if there was some way we could store properties on components that already exist with action graphs, and reference them I would be happy. Even if you don't think this is good, I trust you will make a solution that's good and well thought through.

**Comment by KingOfTheFBI** [comment]
What do you guys suggest for map scenes, then? Custom components do not get uploaded with hammer maps and that seems to be intentional. So we cannot do the workaround of making a custom component (not that having to code for a hammer map is great in the first place).

**Comment by Fortune117** [comment]
I think a component that can dynamically store custom variables might be a good alternative.

**Comment by Metapyziks** [comment]
I'm making a keyvalue component for now, that integrates with the action graph editor so it can drop down list known properties for each object.

**Comment by yuberee** [comment]
Any news on this?
Every day I see some new developer asking how to create variables/properties with ActionGraph.
🙏🏽
